client/session.go

```diff
@@ -1625,74 +1661,163 @@ func (s *session) FetchBlocksFromPeers(
 
 func (s *session) streamBlocksMetadataFromPeers(
 	namespace ident.ID,
-	shard uint32,
-	peers []peer,
+	shardID uint32,
+	peers peers,
 	start, end time.Time,
+	consistencyLevel ReadConsistencyLevel,
 	metadataCh chan<- blocksMetadata,
 	resultOpts result.Options,
 	progress *streamFromPeersMetrics,
 	version FetchBlocksMetadataEndpointVersion,
 ) error {
 	var (
 		wg       sync.WaitGroup
-		errLock  sync.Mutex
-		errLen   int
-		pending  int64
-		multiErr = xerrors.NewMultiError()
+		errLock  sync.RWMutex
+		errors   = make(map[int]error)
+		setError = func(idx int, err error) {
+			errLock.Lock()
+			errors[idx] = err
+			errLock.Unlock()
+		}
+		getErrors = func() []error {
+			var result []error
+			errLock.RLock()
+			for _, err := range errors {
+				if err == nil {
+					continue
+				}
+				result = append(result, err)
+			}
+			errLock.RUnlock()
+			return result
+		}
+		abortError    error
+		setAbortError = func(err error) {
+			errLock.Lock()
+			abortError = err
+			errLock.Unlock()
+		}
+		getAbortError = func() error {
+			errLock.RLock()
+			result := abortError
+			errLock.RUnlock()
+			return result
+		}
+		pending   = int64(len(peers.peers))
+		majority  = atomic.LoadInt32(&s.majority)
```

Comment:
Why do we need an atomic here? In case it changes during the bootstrap?

Reply:
It *could* change during topology change (add replica) but very unlikely.